backend/src/main/scala/bloop/logging/ProcessLogger.scala

```diff
@@ -0,0 +1,87 @@
+package bloop.logging
+
+import java.io.{
+  BufferedReader,
+  ByteArrayOutputStream,
+  InputStream,
+  InputStreamReader,
+  OutputStream,
+  PrintStream
+}
+import scala.annotation.tailrec
+
+/**
+ * Logs the output on stdout and std err of `process` to the `underlying` logger
+ *
+ * @param underlying The logger that receives the output of the process
+ * @param process    The process whose output must be logged.
+ */
+class ProcessLogger(underlying: Logger, process: Process) {
+  private[this] val processOut = process.getInputStream
+  private[this] val processErr = process.getErrorStream
+
+  def start(): Unit = {
+    underlying.debug("Starting to log output from process...")
+    new StreamLogger(underlying.info, processOut).start()
+    new StreamLogger(underlying.error, processErr).start()
+  }
+}
+
+object ProcessLogger {
+
+  private val encoding = sys.props("file.encoding")
+
+  /**
+   * Creates an `OutputStream` from a logging function
+   *
+   * @param logFn A function that logs the given string to a logger.
+   * @return An `OutputStream` that can be used as `System.in` or `System.err`.
+   */
+  def toOutputStream(logFn: String => Unit): OutputStream = {
+    val outputStream = new OutputStream {
+      private val buffer = new ByteArrayOutputStream
+
+      override def write(x: Int): Unit = synchronized {
```

Comment:
Doesn't this mean that if we don't end something with a new line we could be missing logs? Isn't there a way we can reuse some JDK utility instead of implementing our own syncrhonized output stream? Or Guava?

Reply:
The stream will be flushed when there's a newline and when it's closed. That's the best solution I've found to have the logger reproduce the output of the running application.

At first I relied on `autoFlush` to do the right thing, but I'd end up with duplicated newlines when the application did `println()` for instance.

This means that the output of the application may be delayed until it exits or a newline character is seen, but I don't think that we could lose output that way.

Reply:
> This means that the output of the application may be delayed until it exits or a newline character is seen, but I don't think that we could lose output that way.